File: src/services/session_manager.rs
use chrono::{DateTime, Utc};
use dashmap::DashMap;
use serde::Serialize;
use std::sync::Arc;

#[derive(Debug, Clone, Serialize)]
pub struct ActiveSession {
    pub user_id: String,
    pub user_name: String,
    pub device_id: String,
    pub device_name: String,
    pub app_name: String,
    pub app_version: String,
    pub client_ip: String,
    pub last_activity: DateTime<Utc>,
    pub now_playing: Option<NowPlaying>,
}

#[derive(Debug, Clone, Serialize)]
pub struct NowPlaying {
    pub item_id: String,
    pub item_name: String,
    pub item_type: String,
    pub series_name: Option<String>,
    pub runtime_ticks: Option<i64>,
    pub position_ticks: i64,
    pub is_paused: bool,
    pub season_index: Option<i32>,
    pub episode_index: Option<i32>,
    pub primary_image_item_id: Option<String>,
}

#[derive(Clone)]
pub struct SessionManager {
    sessions: Arc<DashMap<String, ActiveSession>>,
}

impl SessionManager {
    pub fn new() -> Self {
        let mgr = Self {
            sessions: Arc::new(DashMap::new()),
        };
        // Spawn cleanup task
        let sessions = mgr.sessions.clone();
        tokio::spawn(async move {
            let mut interval = tokio::time::interval(std::time::Duration::from_secs(30));
            loop {
                interval.tick().await;
                let cutoff = Utc::now() - chrono::Duration::minutes(10);
                sessions.retain(|_, v| v.last_activity > cutoff);
            }
        });
        mgr
    }

    pub fn update_session(
        &self,
        user_id: &str,
        user_name: &str,
        device_id: &str,
        device_name: &str,
        app_name: &str,
        app_version: &str,
        client_ip: &str,
    ) {
        let key = format!("{user_id}:{device_id}");
        self.sessions
            .entry(key)
            .and_modify(|s| {
                s.user_name = user_name.to_string();
                s.device_name = device_name.to_string();
                s.app_name = app_name.to_string();
                if !app_version.is_empty() { s.app_version = app_version.to_string(); }
                if !client_ip.is_empty() { s.client_ip = client_ip.to_string(); }
                s.last_activity = Utc::now();
            })
            .or_insert_with(|| ActiveSession {
                user_id: user_id.to_string(),
                user_name: user_name.to_string(),
                device_id: device_id.to_string(),
                device_name: device_name.to_string(),
                app_name: app_name.to_string(),
                app_version: app_version.to_string(),
                client_ip: client_ip.to_string(),
                last_activity: Utc::now(),
                now_playing: None,
            });
    }

    pub fn set_now_playing(
        &self,
        user_id: &str,
        device_id: &str,
        item_id: &str,
        item_name: &str,
        item_type: &str,
        series_name: Option<&str>,
        runtime_ticks: Option<i64>,
        position_ticks: i64,
        is_paused: bool,
        season_index: Option<i32>,
        episode_index: Option<i32>,
        primary_image_item_id: Option<&str>,
    ) {
        let key = format!("{user_id}:{device_id}");
        if let Some(mut session) = self.sessions.get_mut(&key) {
            session.now_playing = Some(NowPlaying {
                item_id: item_id.to_string(),
                item_name: item_name.to_string(),
                item_type: item_type.to_string(),
                series_name: series_name.map(|s| s.to_string()),
                runtime_ticks,
                position_ticks,
                is_paused,
                season_index,
                episode_index,
                primary_image_item_id: primary_image_item_id.map(|s| s.to_string()),
            });
            session.last_activity = Utc::now();
        }
    }

    pub fn clear_now_playing(&self, user_id: &str, device_id: &str) {
        let key = format!("{user_id}:{device_id}");
        if let Some(mut session) = self.sessions.get_mut(&key) {
            session.now_playing = None;
        }
    }

    pub fn get_active_sessions(&self) -> Vec<ActiveSession> {
        let cutoff = Utc::now() - chrono::Duration::minutes(10);
        self.sessions
            .iter()
            .filter(|entry| entry.value().last_activity > cutoff)
            .map(|entry| entry.value().clone())
            .collect()
    }
}

File: internal/services/session_manager.go
package services

import (
	"sync"
	"time"
)

// nowPlayingTimeout 是 NowPlaying 的存活窗口:超过这段时间没有播放进度刷新
// (SetNowPlaying)就认为该流已经停止。用于防止客户端异常退出未发 Stopped 事件、
// 但仍在浏览导致会话 LastActivity 持续刷新时,僵尸 NowPlaying 一直占用并发流名额。
// 与 handlers 内 activePlaybacks 的 120s 回收阈值保持一致。
const nowPlayingTimeout = 2 * time.Minute

type ActiveSession struct {
	UserID        string      `json:"UserId"`
	UserName      string      `json:"UserName"`
	DeviceID      string      `json:"DeviceId"`
	DeviceName    string      `json:"DeviceName"`
	AppName       string      `json:"AppName"`
	AppVersion    string      `json:"AppVersion"`
	ClientIP      string      `json:"ClientIp"`
	PlaySessionID string      `json:"PlaySessionId,omitempty"`
	LastActivity  time.Time   `json:"LastActivity"`
	NowPlaying    *NowPlaying `json:"NowPlaying,omitempty"`
	// NowPlayingAt 记录 NowPlaying 最近一次被设置/刷新的时间,用于过期判定。
	NowPlayingAt time.Time `json:"-"`
}

type NowPlaying struct {
	ItemID             string  `json:"ItemId"`
	MediaSourceID      string  `json:"MediaSourceId,omitempty"`
	ItemName           string  `json:"ItemName"`
	ItemType           string  `json:"ItemType"`
	SeriesName         *string `json:"SeriesName,omitempty"`
	RuntimeTicks       *int64  `json:"RuntimeTicks,omitempty"`
	PositionTicks      int64   `json:"PositionTicks"`
	IsPaused           bool    `json:"IsPaused"`
	SeasonIndex        *int32  `json:"SeasonIndex,omitempty"`
	EpisodeIndex       *int32  `json:"EpisodeIndex,omitempty"`
	PrimaryImageItemID *string `json:"PrimaryImageItemId,omitempty"`
	PlaySessionID      string  `json:"PlaySessionId,omitempty"`
	PlayMethod         string  `json:"PlayMethod,omitempty"`
}

type SessionManager struct {
	mu       sync.RWMutex
	sessions map[string]*ActiveSession
	stopCh   chan struct{}
}

func NewSessionManager() *SessionManager {
	sm := &SessionManager{
		sessions: make(map[string]*ActiveSession),
		stopCh:   make(chan struct{}),
	}

	go sm.cleanupLoop()
	return sm
}

func (sm *SessionManager) cleanupLoop() {
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			sm.cleanup()
		case <-sm.stopCh:
			return
		}
	}
}

func (sm *SessionManager) cleanup() {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	cutoff := time.Now().Add(-10 * time.Minute)
	playingCutoff := time.Now().Add(-nowPlayingTimeout)
	for k, s := range sm.sessions {
		if s.LastActivity.Before(cutoff) {
			delete(sm.sessions, k)
			continue
		}
		// 会话本身还活跃,但 NowPlaying 已过期(久未刷新进度)→ 清掉播放状态,
		// 释放并发流名额。
		if s.NowPlaying != nil && s.NowPlayingAt.Before(playingCutoff) {
			s.NowPlaying = nil
			s.PlaySessionID = ""
		}
	}
}

func (sm *SessionManager) UpdateSession(userID, userName, deviceID, deviceName, appName, appVersion, clientIP string) {
	key := userID + ":" + deviceID
	sm.mu.Lock()
	defer sm.mu.Unlock()

	if s, ok := sm.sessions[key]; ok {
		s.UserName = userName
		s.DeviceName = deviceName
		s.AppName = appName
		if appVersion != "" {
			s.AppVersion = appVersion
		}
		if clientIP != "" {
			s.ClientIP = clientIP
		}
		s.LastActivity = time.Now()
	} else {
		sm.sessions[key] = &ActiveSession{
			UserID:       userID,
			UserName:     userName,
			DeviceID:     deviceID,
			DeviceName:   deviceName,
			AppName:      appName,
			AppVersion:   appVersion,
			ClientIP:     clientIP,
			LastActivity: time.Now(),
		}
	}
}

func (sm *SessionManager) SetNowPlaying(userID, deviceID string, np *NowPlaying) {
	key := userID + ":" + deviceID
	sm.mu.Lock()
	defer sm.mu.Unlock()

	if s, ok := sm.sessions[key]; ok {
		s.NowPlaying = np
		if np != nil && np.PlaySessionID != "" {
			s.PlaySessionID = np.PlaySessionID
		}
		now := time.Now()
		if np != nil {
			s.NowPlayingAt = now
		}
		s.LastActivity = now
	}
}

func (sm *SessionManager) ClearNowPlaying(userID, deviceID string) {
	key := userID + ":" + deviceID
	sm.mu.Lock()
	defer sm.mu.Unlock()

	if s, ok := sm.sessions[key]; ok {
		s.NowPlaying = nil
		s.PlaySessionID = ""
	}
}

func (sm *SessionManager) ClearNowPlayingBySessionID(sessionID string) bool {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	for _, s := range sm.sessions {
		if sessionMatchesID(s, sessionID) {
			s.NowPlaying = nil
			s.PlaySessionID = ""
			s.LastActivity = time.Now()
			return true
		}
	}
	return false
}

func (sm *SessionManager) HasSession(sessionID string) bool {
	sm.mu.RLock()
	defer sm.mu.RUnlock()

	cutoff := time.Now().Add(-10 * time.Minute)
	for _, s := range sm.sessions {
		if s.LastActivity.After(cutoff) && sessionMatchesID(s, sessionID) {
			return true
		}
	}
	return false
}

func sessionMatchesID(s *ActiveSession, sessionID string) bool {
	if s == nil || sessionID == "" {
		return false
	}
	if s.UserID+"_"+s.DeviceID == sessionID {
		return true
	}
	if s.PlaySessionID == sessionID {
		return true
	}
	return s.NowPlaying != nil && s.NowPlaying.PlaySessionID == sessionID
}

// CountActiveStreams 统计某用户当前真正在播放的并发流数量。只计入 NowPlaying
// 仍新鲜(最近 nowPlayingTimeout 内有进度刷新)且会话活跃的流,读取时即过滤,
// 不依赖后台 cleanup 的 30s 周期,避免僵尸会话误占名额导致错误拦截。
func (sm *SessionManager) CountActiveStreams(userID string) int {
	sm.mu.RLock()
	defer sm.mu.RUnlock()

	now := time.Now()
	sessionCutoff := now.Add(-10 * time.Minute)
	playingCutoff := now.Add(-nowPlayingTimeout)
	n := 0
	for _, s := range sm.sessions {
		if s.UserID != userID || s.NowPlaying == nil {
			continue
		}
		if s.LastActivity.After(sessionCutoff) && s.NowPlayingAt.After(playingCutoff) {
			n++
		}
	}
	return n
}

func (sm *SessionManager) GetActiveSessions() []ActiveSession {
	sm.mu.RLock()
	defer sm.mu.RUnlock()

	cutoff := time.Now().Add(-10 * time.Minute)
	var result []ActiveSession
	for _, s := range sm.sessions {
		if s.LastActivity.After(cutoff) {
			result = append(result, *s)
		}
	}
	return result
}

func (sm *SessionManager) Stop() {
	close(sm.stopCh)
}
